src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,183 @@
         }
 
-        private static bool IsPathInTreeRootedAtSolutionDirectory(string path)
+        private void AddProject(SolutionModel solution, string pathRelativeToSolutionFile, string fullPath, SolutionFolderModel solutionFolder)
         {
-            return !path.StartsWith("..");
-        }
+            // Open project instance to see if it is a valid project
+            ProjectRootElement projectRootElement = ProjectRootElement.Open(fullPath);
+            SolutionProjectModel project;
+            try
+            {
+                project = solution.AddProject(pathRelativeToSolutionFile, null, solutionFolder);
+            }
+            catch (SolutionArgumentException ex) when (ex.ParamName == "projectTypeName")
+            {
+                // If guid is not identified by vs-solutionpersistence, check in project element itself
+                var guid = projectRootElement.GetProjectTypeGuid();
+                if (string.IsNullOrEmpty(guid))
+                {
+                    Reporter.Error.WriteLine(CommonLocalizableStrings.UnsupportedProjectType, fullPath);
+                    return;
+                }
+                project = solution.AddProject(pathRelativeToSolutionFile, guid, solutionFolder);
+            }
+            // Generate intermediate solution folders
+            if (solutionFolder is null && !_inRoot)
+            {
+                var relativePathDirectory = Path.GetDirectoryName(pathRelativeToSolutionFile);
+                if (!string.IsNullOrEmpty(relativePathDirectory))
+                {
+                    SolutionFolderModel relativeSolutionFolder = solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(relativePathDirectory));
+                    project.MoveToFolder(relativeSolutionFolder);
+                    // Avoid duplicate folder/project names
+                    if (project.Parent is not null && project.Parent.ActualDisplayName == project.ActualDisplayName)
+                    {
+                        solution.RemoveFolder(project.Parent);
+                    }
+                }                
+            }
+            // Add settings based on existing project instance
+            ProjectInstance projectInstance = new ProjectInstance(projectRootElement);
+            string projectInstanceId = projectInstance.GetProjectId();
+            if (!string.IsNullOrEmpty(projectInstanceId))
+            {
+                project.Id = new Guid(projectInstanceId);
+            }
+            var projectInstanceBuildTypes = projectInstance.GetConfigurations();
+            var projectInstancePlatforms = projectInstance.GetPlatforms();
 
-        private static string TrimProject(string path)
-        {
-            return Path.GetDirectoryName(path);
-        }
+            foreach (var solutionPlatform in solution.Platforms)
+            {
+                var projectPlatform = projectInstancePlatforms.FirstOrDefault(
```

Comment:
If there are multiple, we only care about the first?

Reply:
I agree. Here I tried to avoid breaking any tests and changing the behavior of the command as to not impact users. 
Here is the current test: https://github.com/dotnet/sdk/blob/e74e404c24adbcea8468529191a950d86c5473c7/test/dotnet-sln.Tests/GivenDotnetSlnAdd.cs#L1153

Reply:
I also think if it needs to map to a build type/platform that does not match the ones in the solution, it makes sense to default to the first one